### ISSUE TITLE
More robust LoadFromCollection support for Dictionaries and/or dynamic ExpandoObject

### ISSUE THREAD

**Issue by Awsmolak** [EPPlusSoftware/EPPlus#147]
I have some dynamic classes that I need to quickly dump out to worksheets.

If I use a list of Dictionaries (Dictionary<string, object>), LoadFromCollection throws a "System.Reflection.TargetParameterCountException".

If I use a list of ExpandoObjects, LoadFromCollection loads the rows, but only one column which contains the first property (key) and it's value. (ex. `[myProperty, value]')

In order to make sure that I was not misunderstanding, I implemented a class for testing to store the same properties/values as the ExpandoObject and Dictionary and it works as expected. In that class I defined my properties as object type. So structurally, all three implementations are almost identical! However, I will not know my properties at compile-time so that approach is obviously unusable. 

Any advice here?

**Comment by Awsmolak** [comment]
Here is example demonstration code

```csharp
using System;
using System.Collections.Generic;
using System.Dynamic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using OfficeOpenXml;

namespace Test
{
    public static class EppDynamicTest
    {

        public static void Test()
        {
            var ids = new List<string>(){"A","B","C","D","E" };

            var rnd = new Random();

            var dat1 = ids.Select(id => rnd.NextDouble());
            var dat2 = ids.Select(id => rnd.NextDouble());

            WriteToExcel(ids,dat1.ToList(),dat2.ToList());
        }

        public static void WriteToExcel(List<string> ids, List<double> col1Data, List<double> col2Data)
        {
            var Test1 = new List<Result>();
            var Test2 = new List<ExpandoObject>();
            var Test3 = new List<Dictionary<string, object>>();
            
            

            for (int i = 0; i < ids.Count(); i++)
            {
                Test1.Add(new Result() { Id = ids[i], D1 = col1Data[i], D2 = col2Data[i] });

                dynamic eo = new ExpandoObject();
                eo.Id = ids[i];
                eo.D1 = ids[i];
                eo.D2 = ids[i];
                Test2.Add(eo);

                var dict = new Dictionary<string, object>();
                dict["Id"] = ids[i];
                dict["D1"] = col1Data;
                dict["D2"] = col2Data;
                Test3.Add(dict);
            }

            var outputPackage = new ExcelPackage();

            outputPackage.Workbook.Worksheets.Add("Test1");
            outputPackage.Workbook.Worksheets.Add("Test2");
            outputPackage.Workbook.Worksheets.Add("Test3");

            var sheet = outputPackage.Workbook.Worksheets["Test1"];

            //this works as intended
            sheet.Cells["B2"].LoadFromCollection(Test1, true);

            sheet = outputPackage.Workbook.Worksheets["Test2"];

            //this doesn't
            sheet.Cells["B2"].LoadFromCollection(Test2, true);

            sheet = outputPackage.Workbook.Worksheets["Test3"];

            //this throws "System.Reflection.TargetParameterCountException"
            //sheet.Cells["B2"].LoadFromCollection(Test3, true);

            outputPackage.SaveAs(new FileInfo("C:\\temp\\test.xlsx"));
        }


        public class Result
        {
            public string Id { get; set; }
            public object D1 { get; set; }
            public object D2 { get; set; }
        }
    }
}
```

**Comment by JanKallman** [comment]
LoadFromCollection currently only supports reflection, but this is an excellent enhancment. We'll look into it.

**Comment by Awsmolak** [comment]
I sidestepped this limitation by using the following extension methods to transform either a List of Dictionary<string,object> or ExpandoObject to DataTable. I haven't extensively tested it, but it seems to work for both cases so far. 

```csharp
public static class DataTableExtensions
    {
        /// <summary>
        /// Create a datatable from a list of <see cref="IEnumerable{T}"/>
        /// </summary>
        /// <param name="data">The list can be created from a dictionary with Dictionary.Values.ToList()</param>
        /// <param name="tableName">Name of the data table</param>
        /// <returns></returns>
        public static DataTable ToDataTable(this IEnumerable<dynamic> data, string tableName)
        {
            return data.ToList().ToDataTable(tableName);
        }

        /// <summary>
        /// Create a datatable from a list of ExpandoObjects
        /// </summary>
        /// <param name="list">The list can be created from a dictionary with Dictionary.Values.ToList()</param>
        /// <param name="tableName">Name of the data table</param>
        /// <returns></returns>
        public static DataTable ToDataTable(this List<ExpandoObject> list, string tableName)
        {

            if (list == null || list.Count == 0)
            {
                return null;
            }

            //build columns
            var props = (IDictionary<string, object>)list[0];
            var t = new DataTable(tableName);
            foreach (var prop in props)
            {
                t.Columns.Add(new DataColumn(prop.Key, prop.Value.GetType()));
            }
            //add rows
            foreach (var row in list)
            {
                var data = t.NewRow();
                foreach (var prop in (IDictionary<string, object>)row)
                {
                    data[prop.Key] = prop.Value;
                }
                t.Rows.Add(data);
            }
            return t;
        }

        /// <summary>
        /// Create a datatable from a list of ExpandoObjects
        /// </summary>
        /// <param name="list">The list can be created from a dictionary with Dictionary.Values.ToList()</param>
        /// <param name="tableName">Name of the data table</param>
        /// <returns></returns>
        public static DataTable ToDataTable(this List<Dictionary<string,object>> list, string tableName)
        {

            if (list == null || list.Count == 0)
            {
                return null;
            }

            //build columns
            var props = list[0];
            var t = new DataTable(tableName);
            foreach (var prop in props)
            {
                t.Columns.Add(new DataColumn(prop.Key, prop.Value.GetType()));
            }
            //add rows
            foreach (var row in list)
            {
                var data = t.NewRow();
                foreach (var prop in row)
                {
                    data[prop.Key] = prop.Value;
                }
                t.Rows.Add(data);
            }
            return t;
        }
    }
```

**Comment by swmal** [comment]
I have started work on this here: https://github.com/EPPlusSoftware/EPPlus/tree/feature/load_functions

Will implemented LoadFromDictionaries and LoadFromExpandoObjects as two new methods on ExcelRangeBase since they differ a bit from the reflectionbased LoadFromCollection.

**Comment by swmal** [comment]
New method LoadFromDictionaries available in the latest version 5.2.1.
